File: solutions/fortran/hamming/1/hamming.f90
module hamming
  implicit none
contains

  function compute(strand1, strand2, distance)
      character(*) :: strand1, strand2
      integer :: distance
      integer :: n
      logical :: compute

      if (len(strand1) /= len(strand2)) then
        compute = .false.
        distance = 0
      else
        compute = .true.
        distance = count([(strand1(n:n) /= strand2(n:n), n = 1, len(strand1))])
      end if
  end function compute

end module hamming

File: solutions/c/hamming/1/hamming.c
#include "hamming.h"

static int compute_r(const char *lhs, const char *rhs, int diff) {
  if (*lhs == '\0' && *rhs == '\0') {
    return diff;
  }
  if (*lhs == '\0' || *rhs == '\0') {
    return -1;
  }
  if (*lhs == *rhs) {
    return compute_r(lhs + 1, rhs + 1, diff);
  }
  return compute_r(lhs + 1, rhs + 1, diff + 1);
}

int compute(const char *lhs, const char *rhs) { return compute_r(lhs, rhs, 0); }
